include/app_event_manager/app_event_manager.h

```diff
@@ -0,0 +1,422 @@
+ *
+ * @param hook_fn Hook function.
+ */
+#define APP_EVENT_HOOK_PREPROCESS_REGISTER_FIRST(hook_fn)                      \
+	const struct {} __event_hook_preprocess_first_sub_redefined = {};  \
+	_APP_EVENT_HOOK_PREPROCESS_REGISTER(hook_fn, _EM_MARKER_FIRST_ELEMENT)
+
+/**
+ * @brief Register event hook on the start of event processing.
+ *
+ * The event hook called when the event is being processed.
+ * The hook function should have a form `void hook(const struct app_event_header *aeh)`.
+ *
+ * @param hook_fn Hook function.
+ */
+#define APP_EVENT_HOOK_PREPROCESS_REGISTER(hook_fn) \
+	_APP_EVENT_HOOK_PREPROCESS_REGISTER(hook_fn, _EM_SUBS_PRIO_ID(_EM_SUBS_PRIO_NORMAL))
+
+/**
+ * @brief Register event hook on the start of event processing. The hook would be called last.
+ *
+ * The event hook called when the event is being processed.
+ * The hook function should have a form `void hook(const struct app_event_header *aeh)`.
+ * The macro makes sure that the hook provided here is called last.
+ * Only one hook can be registered with this macro.
+ *
+ * @param hook_fn Hook function.
+ */
+#define APP_EVENT_HOOK_PREPROCESS_REGISTER_LAST(hook_fn)                      \
+	const struct {} __event_hook_preprocess_last_sub_redefined = {};  \
+	_APP_EVENT_HOOK_PREPROCESS_REGISTER(hook_fn, _EM_MARKER_FINAL_ELEMENT)
+
+/**
+ * @brief Register event hook on the end of event processing. The hook would be called first.
+ *
+ * The event hook called after the event is processed.
+ * The hook function should have a form `void hook(const struct app_event_header *aeh)`.
+ * The macro makes sure that the hook provided here is called first.
+ * Only one hook can be registered with this macro.
+ *
+ * @param hook_fn Hook function.
+ */
+#define APP_EVENT_HOOK_POSTPROCESS_REGISTER_FIRST(hook_fn)                      \
+	const struct {} __event_hook_postprocess_first_sub_redefined = {};  \
+	_APP_EVENT_HOOK_POSTPROCESS_REGISTER(hook_fn, _EM_MARKER_FIRST_ELEMENT)
+
+/**
+ * @brief Register event hook on the end of event processing.
+ *
+ * The event hook called after the event is processed.
+ * The hook function should have a form `void hook(const struct app_event_header *aeh)`.
+ *
+ * @param hook_fn Hook function.
+ */
+#define APP_EVENT_HOOK_POSTPROCESS_REGISTER(hook_fn)	\
+	_APP_EVENT_HOOK_POSTPROCESS_REGISTER(hook_fn,	\
+	_EM_SUBS_PRIO_ID(_EM_SUBS_PRIO_NORMAL))
+
+/**
+ * @brief Register event hook on the end of event processing. The hook would be called last.
+ *
+ * The event hook called after the event is processed.
+ * The hook function should have a form `void hook(const struct app_event_header *aeh)`.
+ * The macro makes sure that the hook provided here is called last.
+ * Only one hook can be registered with this macro.
+ *
+ * @param hook_fn Hook function.
+ */
+#define APP_EVENT_HOOK_POSTPROCESS_REGISTER_LAST(hook_fn)                      \
+	const struct {} __event_hook_postprocess_last_sub_redefined = {};  \
```

Comment:
what is the purpose of this variable?

Reply:
It is a marker indicating that only one hook can be registered as Last.